### PR TITLE
Updated camera

### PR DESCRIPTION
**Pull request by nickwalton**
Adds Holodeck Camera class which can enable any number of agents to have cameras. Optimizations have been made to increase the speed from previous attempts by bypassing the "ReadPixels" function and reading from the GPU without performing a gpu flush.

### CODE REVIEW

File: Source/Holodeck/General/Private/HolodeckViewportClient.cpp [hunk 1]
```diff
@@ -16,8 +17,19 @@ void UHolodeckViewportClient::HolodeckTakeScreenShot() {
 void UHolodeckViewportClient::Draw(FViewport * ViewportParam, FCanvas * SceneCanvas) {
 	Super::Draw(ViewportParam, SceneCanvas);
 	HolodeckTakeScreenShot();
+	UE_LOG(LogHolodeck, Log, TEXT("@@@@@ Draw called"));
```
Comment (mitchneutron) [inline Source/Holodeck/General/Private/HolodeckViewportClient.cpp:20]: We should take out these debug statements so they don't clutter up the log.

File: Source/Holodeck/General/Private/HolodeckViewportClient.cpp [hunk 2]
```diff
@@ -16,8 +17,19 @@ void UHolodeckViewportClient::HolodeckTakeScreenShot() {
...
+	for (auto& Cam : Cameras) {
+		Cam.Value->Capture();
+	}
+
+	UE_LOG(LogHolodeck, Log, TEXT("@@@@@ Draw Ended"));
```
Comment (mitchneutron) [inline Source/Holodeck/General/Private/HolodeckViewportClient.cpp:25]: This one too

File: Source/Holodeck/General/Private/HolodeckViewportClient.cpp (hunk 1 above)
Comment (mitchneutron) [reply Source/Holodeck/General/Private/HolodeckViewportClient.cpp:20]: Also, considering the workaround you did with not flushing the rendering thread, is it still necessary for us to capture the pixels at the time that the viewport renders?

File: Source/Holodeck/General/Private/RenderRequest.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	// Queue up the task of rendering the scene in the render thread
+	TGraphTask<FRenderRequest>::CreateTask().ConstructAndDispatchWhenReady(*this);
+}
+
+void FRenderRequest::ExecuteTask()
+{
+	TArray<FColor> SurfaceData;
+	FRHICommandListImmediate& RHICmdList = GetImmediateCommandList_ForRenderCommand();
+	FTextureRenderTargetResource*  rt_resource = TargetTexture->GetRenderTargetResource();
```
Comment (mitchneutron) [inline Source/Holodeck/General/Private/RenderRequest.cpp:17]: The variable "rt_resource" should follow the name convention that we have. RTResource maybe?

File: Source/Holodeck/General/Private/RenderRequest.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+	if (rt_resource != nullptr) {
+		const FTexture2DRHIRef& rhi_texture = rt_resource->GetRenderTargetTexture();
+		FIntPoint size;
```
Comment (mitchneutron) [inline Source/Holodeck/General/Private/RenderRequest.cpp:21]: These too. Variables need to be capitalized.

File: Source/Holodeck/General/Public/HolodeckViewportClient.h [hunk 5]
```diff
@@ -47,6 +47,12 @@ class HOLODECK_API UHolodeckViewportClient : public UGameViewportClient
 	  */
 	bool BufferIsSet() { return Buffer != nullptr; };
 
+	/**
```
Comment (mitchneutron) [inline Source/Holodeck/General/Public/HolodeckViewportClient.h:50]: We should make sure to document the public functions.

File: Source/Holodeck/General/Public/HolodeckViewportClient.h [hunk 6]
```diff
@@ -55,6 +61,9 @@ class HOLODECK_API UHolodeckViewportClient : public UGameViewportClient
 	  */
 	void HolodeckTakeScreenShot();
 
+	//void ClearNullCameras();
```
Comment (mitchneutron) [inline Source/Holodeck/General/Public/HolodeckViewportClient.h:64]: Clear out commented code

File: Source/Holodeck/General/Public/RenderRequest.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "Holodeck.h"
+
+#include "CoreMinimal.h"
+#include "Engine/TextureRenderTarget2D.h"
+
+// Adapted from Microsoft Airsim Open Source Project
```
Comment (mitchneutron) [inline Source/Holodeck/General/Public/RenderRequest.h:8]: I'm glad that you sited the source of inspiration!

File: Source/Holodeck/General/Public/RenderRequest.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+		UTextureRenderTarget2D* TargetTexture;
+
+	public:
+
+		/**
+		* Retrieves the rendered texture from the GPU without flushing the GPU like ReadPixels() does. 
+		*/
+		virtual void RetrievePixels(FColor* Buffer, UTextureRenderTarget2D* TargetTexture);
+
+		void DoTask(ENamedThreads::Type CurrentThread, const FGraphEventRef& MyCompletionGraphEvent)
```
Comment (mitchneutron) [inline Source/Holodeck/General/Public/RenderRequest.h:25]: The code is very self documenting, which is good. We should still make sure to document the public functions though.

File: Source/Holodeck/Sensors/Public/HolodeckCamera.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	UPROPERTY(EditAnywhere)
+	int CaptureHeight = 256;
+
+private:
+
+	bool bPointerGivenToViewport = false;
+	UHolodeckViewportClient* ViewportClient;
+	FColor* Buffer;
+	FRenderRequest RenderRequest;
+};
```
Comment (mitchneutron) [inline Source/Holodeck/Sensors/Public/HolodeckCamera.h:58]: Newline here

**Review by mitchneutron — CHANGES_REQUESTED** [review]
As far as your logic and actual code, it looks great. The workaround is awesome. The only architecture concern I have is with the  camera's capture being called from the viewport object instead of just in the tick function. I'm unsure if it's still necessary (it might be!) with the workaround you did. It would be best if we could avoid that dependency. 

Other than that, just nits for the style guide. Maybe we can talk about loosening it up a little? I think it does have value to strictly follow it though.

**Comment by nickwalton** [comment]
Realized this branch is slightly corrected. Transferred changes to new branch and will create pull request there.

**Pull request closed without merging**